### PR TITLE
build: update toolchain to nightly-2024-04-18

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

- Allows `clippy.readonly_write_lock = "allow"`. We need to check it afterward.
- Remove some unused structs

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/table/src/predicate.rs [hunk 1]
```diff
@@ -573,6 +573,7 @@ mod tests {
         let file = std::fs::OpenOptions::new()
             .write(true)
             .create(true)
+            .truncate(true)
```
Comment (tisonkun) [inline src/table/src/predicate.rs:576]: Seems unrelated change?

File: src/query/src/tests/time_range_filter_test.rs [hunk 2]
```diff
@@ -58,7 +58,7 @@ struct DataSourceWrapper {
 
 impl DataSource for DataSourceWrapper {
     fn get_stream(&self, request: ScanRequest) -> Result<SendableRecordBatchStream, BoxedError> {
-        *self.filter.write().unwrap() = request.filters.clone();
+        self.filter.write().unwrap().clone_from(&request.filters);
```
Comment (tisonkun) [inline src/query/src/tests/time_range_filter_test.rs:61]: I guess this comes from clippy, but why `clone_from`/`clone_into` preferred?

File: src/common/time/src/timestamp.rs [hunk 3]
```diff
@@ -381,7 +381,7 @@ impl Timestamp {
 
     pub fn from_chrono_datetime(ndt: NaiveDateTime) -> Option<Self> {
         let sec = ndt.and_utc().timestamp();
-        let nsec = ndt.timestamp_subsec_nanos();
+        let nsec = ndt.and_utc().timestamp_subsec_nanos();
```
Comment (tisonkun) [inline src/common/time/src/timestamp.rs:384]: This seems to change the semantic. Why do we need it?

File: src/common/procedure/src/local/rwlock.rs [hunk 4]
```diff
@@ -18,6 +18,8 @@ use std::sync::{Arc, Mutex};
 
 use tokio::sync::{OwnedRwLockReadGuard, OwnedRwLockWriteGuard, RwLock};
 
+// Override warnings that those fields are not accessed.
+#[allow(dead_code)]
```
Comment (tisonkun) [inline src/common/procedure/src/local/rwlock.rs:22]: Is here a `TODO` to drop this suppression?

File: src/common/meta/src/state_store.rs [hunk 5]
```diff
@@ -79,6 +79,8 @@ fn decode_kv(kv: KeyValue) -> Result<(String, Vec<u8>)> {
     Ok((key, value))
 }
 
+// Override warnings that those fields are not accessed.
+#[allow(dead_code)]
```
Comment (tisonkun) [inline src/common/meta/src/state_store.rs:83]: Is here a TODO to drop this suppression?

File: src/common/meta/src/sequence.rs [hunk 6]
```diff
@@ -164,15 +164,15 @@ impl Inner {
 
             let req = CompareAndPutRequest {
                 key: key.to_vec(),
-                expect,
+                expect: expect.clone(),
```
Comment (tisonkun) [inline src/common/meta/src/sequence.rs:167]: This seems to introduce an extra clone?

**Review by tisonkun — COMMENTED** [review]
Rest LGTM. Great patch!

File: src/common/meta/src/sequence.rs (hunk 6 above)
Comment (waynexia) [reply src/common/meta/src/sequence.rs:167]: I don't know why it didn't need before

File: src/common/procedure/src/local/rwlock.rs (hunk 4 above)
Comment (waynexia) [reply src/common/procedure/src/local/rwlock.rs:22]: This doesn't look like a TODO. How would you like to proceed?

File: src/common/time/src/timestamp.rs (hunk 3 above)
Comment (waynexia) [reply src/common/time/src/timestamp.rs:384]: it's deprecated

File: src/mito2/src/cache/cache_size.rs [hunk 7]
```diff
@@ -137,6 +137,6 @@ mod tests {
     fn test_parquet_meta_size() {
         let metadata = parquet_meta();
 
-        assert_eq!(956, parquet_meta_size(&metadata));
+        assert_eq!(964, parquet_meta_size(&metadata));
```
[the remark this answers is not in the document]
Comment (waynexia) [reply src/mito2/src/cache/cache_size.rs:140]: Maybe some indirect dep changes its implement? IDK

File: src/table/src/predicate.rs (hunk 1 above)
Comment (waynexia) [reply src/table/src/predicate.rs:576]: We need to truncate here.

Comment (waynexia) [reply src/table/src/predicate.rs:576]: https://rust-lang.github.io/rust-clippy/master/index.html#/suspicious_open_options

File: src/common/meta/src/sequence.rs (hunk 6 above)
Comment (tisonkun) [reply src/common/meta/src/sequence.rs:167]: This is because it's a retry loop. If the request failed, we always update `expect` so NLL analysis should know here is no need for cloning.

If it can't compile, either the `clone_from` clippy has an edge case, or it's a compiler regression.

File: src/table/src/predicate.rs (hunk 1 above)
Comment (tisonkun) [reply src/table/src/predicate.rs:576]: Got.

File: src/mito2/src/cache/cache_size.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (tisonkun) [reply src/mito2/src/cache/cache_size.rs:140]: Just wonder if it would be a breaking data format change ...

File: src/common/procedure/src/local/rwlock.rs (hunk 4 above)
Comment (tisonkun) [reply src/common/procedure/src/local/rwlock.rs]: Pushed b7a57b38f85953d07d4085cf1b3d43986a224e6e to resolve this.

File: src/common/meta/src/state_store.rs (hunk 5 above)
Comment (tisonkun) [reply src/common/meta/src/state_store.rs]: Pushed b7a57b38f85953d07d4085cf1b3d43986a224e6e to resolve this.

File: src/common/meta/src/state_store.rs [hunk 8]
```diff
@@ -79,19 +80,21 @@ fn decode_kv(kv: KeyValue) -> Result<(String, Vec<u8>)> {
...
 }
 
-fn split_value(value: &[u8], max_value_size: Option<usize>) -> SplitValue<'_> {
+fn split_value(value: Vec<u8>, max_value_size: Option<usize>) -> SplitValue {
     if let Some(max_value_size) = max_value_size {
         if value.len() <= max_value_size {
             SplitValue::Single(value)
         } else {
-            SplitValue::Multiple(value.chunks(max_value_size).collect::<Vec<_>>())
+            let mut values = vec![];
```
Comment (tisonkun) [inline src/common/meta/src/state_store.rs:93]: This also reduce a clone. Since previously:

```rust
    pub fn with_value(mut self, value: impl Into<Vec<u8>>) -> Self {
        self.value = value.into();
        self
    }
```

would clone the inner value chunk.

File: src/common/meta/src/sequence.rs (hunk 6 above)
Comment (tisonkun) [reply src/common/meta/src/sequence.rs]: Resolve in 9a8bcac12b6f1e68c899ddcf88187d216262228e.

File: src/mito2/src/cache/cache_size.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (tisonkun) [reply src/mito2/src/cache/cache_size.rs:140]: Resolved as we have parquet read/write test.

**Review by tisonkun — APPROVED** [review]
LGTM.

File: Cargo.toml [hunk 9]
```diff
@@ -70,7 +70,10 @@ license = "Apache-2.0"
 clippy.print_stdout = "warn"
 clippy.print_stderr = "warn"
 clippy.implicit_clone = "warn"
+clippy.readonly_write_lock = "allow"
```
Comment (tisonkun) [inline Cargo.toml:73]: This should be a todo that we later revisit if locks can improved. And we should suppress the warning in place where it's not the case.

We can create an issue for it.

**Comment by tisonkun** [comment]
cc @MichaelScofield @evenyag @v0y4g3r @zhongzc PTAL.

**Review by MichaelScofield — APPROVED** [review]

File: src/common/meta/src/state_store.rs [hunk 10]
```diff
@@ -79,17 +80,21 @@ fn decode_kv(kv: KeyValue) -> Result<(String, Vec<u8>)> {
...
+            for chunk in value.into_iter().chunks(max_value_size).into_iter() {
```
Comment (evenyag) [inline src/common/meta/src/state_store.rs:94]: @WenyXu Will this introduce additional allocations? The original `SplitValue` holds `Vec<&[u8]>`.

Comment (tisonkun) [reply src/common/meta/src/state_store.rs:94]: No. This call into_iter. Actually the previous code has an extra alloc. See LibraryLoader.

Comment (tisonkun) [reply src/common/meta/src/state_store.rs:94]: We move the bytes here. Previously it uses a ref but we can actually take the ownership. This calls delegate to `itertools` `IntoChunks`, not the slice's `chunks` that may clone.
